jbi/retry.py

```diff
@@ -0,0 +1,61 @@
+import asyncio
+import logging
+from datetime import datetime, timedelta
+from os import getenv
+from time import sleep
+from typing import AsyncIterator
+
+import jbi.runner as runner
+from jbi.configuration import ACTIONS
+from jbi.queue import QueueItem, get_dl_queue
+
+CONSTANT_RETRY = getenv("CONSTANT_RETRY", "false") == "true"
+RETRY_TIMEOUT_DAYS = getenv("RETRY_TIMEOUT_DAYS", 7)
+
+
+async def retry_failed():
+    queue = get_dl_queue()
+    logger = logging.getLogger(__name__)
+    min_event_timestamp = datetime.now() - timedelta(days=int(RETRY_TIMEOUT_DAYS))
+
+    # load all bugs from DLQ
+    bugs: dict[int, AsyncIterator[QueueItem]] = await queue.retrieve()
+
+    for bugid, items in bugs.items():
+        prev_failed = False
+        async for item in items:
+            # skip if any previous retries for this bug have already failed
+            if prev_failed:
+                logger.info(
+                    "skipping event %s - previous events have failed for this bug",
+                    item.identifier,
+                )
+                continue
+
+            # skip and delete item if we have exceeded max_timeout
+            if item.timestamp < min_event_timestamp:
+                logger.warn("removing expired event %s", item.identifier)
+                await queue.done(item)
+                continue
+
+            try:
+                runner.execute_action(item.payload, ACTIONS)
+                await queue.done(item)
+            except Exception as ex:
+                # write well formed log that could be alerted on
+                logger.error(
+                    "failed to reprocess event %s. error: %s", item.identifier, ex
+                )
+                prev_failed = True
```

Comment:
Instead of using a `prev_failed` sentinel, could we `break` here if we encounter an exception? This will take us to the next bug in the queue. If we want to keep the info log, we could do that before we break and say something like `logger.info(skipping remaining events for bug %s, bugid)`

Reply:
That is how I had the code originally as it's definitely more efficient. But then we don't log anything for the events we're skipping. I started to write something that would log "skipping X future events for bug Y", but that seemed like a truncation of data. So then I went with this.

This also helped with the summary metrics we're logging at the end of the run.

With all of that said, is there a reasonable way to break early and still produce meaningful logs and metrics?

Reply:
> "skipping X future events for bug Y"

I think this could be enough, but also maybe this means we should add a 

```python
def list(bug_id: Optional[int] = None)
    ...
```

method so that we can return the identifiers in the queue without trying to load them. That way, we could do something like:

```python
try:
    runner.execute_action(item.payload, ACTIONS)
    metrics["events_processed"] += 1
    await queue.done(item)
except Exception as ex:
    logger.exception(
        "failed to reprocess event %s. error: %s", item.identifier, ex
    )
    metrics["events_failed"] += 1
    skipped_items = await queue.list(bugid)
    logger.info(
        "skipping events %s - previous events have failed for this bug",
        ",".join(skipped_items),
    )
    metrics["events_skipped"] += len(skipped_items)
```

Reply:
[63057fc](https://github.com/mozilla/jira-bugzilla-integration/pull/918/commits/63057fc3b3be364f6a73fed1428d02997551cfac)